### ISSUE TITLE
Automatically create knobs based on PropTypes

### ISSUE THREAD

**Issue by lucasconstantino** [storybook-eol/storybook-addon-knobs#63]
It would be awesome if this addon could automatically create knobs based on propTypes. [Atellier](https://github.com/scup/atellier) does this kind of work already, but the project's development looks frozen.

Here goes a work in progress project based on Atellier: https://github.com/lucasconstantino/storybook-addon-smart-knobs

**Comment by arunoda** [comment]
@lucasconstantino This is great.
Actually, we are working on a propType integration to storybook directly.
See this babel plugin: https://github.com/kadirahq/babel-plugin-react-docgen

We are going to ship this into Storybook with the next minor version.

Basically, with that you can get propTypes and any other react-docgen support into for any React Class.

I hope you could use that as well.
Once we out that, I think it's pretty great to do this on storybook knobs itself.
We are pretty happy to get your help on that.

**Comment by lucasconstantino** [comment]
Ok, that's really good news. Once this is out I'll probably first refactor my addon to make it a POC a then proceed to create a pull-request to Knobs or something like that.

**Comment by arunoda** [comment]
@lucasconstantino that's a great plan.

**Comment by moimael** [comment]
Any progress on this one ? :)

**Comment by lucasconstantino** [comment]
From me, not really. [storybook-addon-smart-knobs](https://github.com/lucasconstantino/storybook-addon-smart-knobs) is already usable though, feel free to give it a try.

**Comment by moimael** [comment]
It seems to clash with the info decorator though. I can see the props of the decorator but not the displayed component's props.

**Comment by lucasconstantino** [comment]
@moimael indeed the way this is done now will not properly work with other components. Actually, it might work if you control the order this plugin is set.

Basically there is no way to properly identify the component's properties "automatically" in a HOC chain, so to address this issue what this plugin does is render the component and use the rendered output as source of information on the proptypes. Yech!

@arunoda and his team developed a babel plugin to attach react-docgen information to the components. This is supposedly the solution for the problem described above, but I had no time yet to try this out.

@moimael any chance you could post your code here? I would gladly try to help you figure it out while no better solution is out there.

**Comment by joevbruno** [comment]
I started to do formulate my own implementation of 'smart' knobs before I found @lucasconstantino project. However, in both that implementation and my own, I am running into an issue when attempting to dynamically `add` multiple components as stories. I want to do this:

```
const stories = storiesOf("Controls", module)
  .addDecorator(withSmartKnobs)
  .addDecorator(withKnobs);

// controls is just a large object of globally shared components available as a webpack alias called "Controls"
// i.e. import * as controls from "Controls";
Object.keys(controls).forEach((componentName) => {
    const Component = controls[componentName];

    stories.add(`Controls.${componentName}`, () => (<Component />));
});
```

but the result is either that 1) the Knobs don't show up at all or 2) I get a `cannot get property null of undefined' as referenced [here](https://github.com/storybooks/storybook-addon-knobs/issues/31) in another Knobs issue. This appears to be a race condition issue with the Knobs store not being setup yet (at times). Also, occasionally, the first component will end up with ALL of the Knobs, while the other components have none. Any help would be greatly appreciated.
